### ISSUE TITLE
Bottled Python 3.5.2 not able to access OSX Keychain via ctypes

### ISSUE THREAD

**Issue by n8henrie** [Homebrew/brew#594]
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?

- 2015 MBA i7 on 10.11.6
- Homebrew 0.9.9 (git revision 1455a; last commit 2016-07-24)
- Homebrew/homebrew-core (git revision 0c8b; last commit 2016-07-26)

### Bug reports:

Bottled python 3.5.2 not able to access OSX Keychain.

tl;dr at the bottom

I've been going bonkers for a couple months now trying to figure out jaraco/keyring#219 -- basically, OS X gives me error [25293 (errSecAuthFailed)](https://developer.apple.com/library/mac/documentation/Security/Reference/keychainservices/#//apple_ref/c/econst/errSecAuthFailed) when I try to access my primary login keychain via ctypes, *specifically with the current Python 3.5.2 bottle*. As far as I can tell, everything else with this Python is working fine, just the keychain access is not working.

Here's my example script, which looks for a generic application password I have created for testing:

- service `keyring_test`
- account `foo`
- password `bar`

The script runs perfectly in [numerous scenarios](https://github.com/jaraco/keyring/issues/219#issuecomment-233054095) including other versions of Python2 and Python3, installed with homebrew or with [pyenv](https://github.com/yyuu/pyenv), including Python 3.5.2 installed with pyenv.

```python3
"""keychain_ctypes.py :: Test script to access keychain from ctypes."""
import ctypes
import sys

sec_path = "/System/Library/Frameworks/Security.framework/Versions/A/Security"
kc = ctypes.CDLL(sec_path).SecKeychainFindGenericPassword

kc.argtypes = (
    ctypes.c_void_p,
    ctypes.c_uint32,
    ctypes.c_char_p,
    ctypes.c_uint32,
    ctypes.c_char_p,
    ctypes.POINTER(ctypes.c_uint32),  # passwordLength
    ctypes.POINTER(ctypes.c_void_p),  # passwordData
    ctypes.POINTER(ctypes.c_void_p),  # itemRef
)

length = ctypes.c_uint32()
data = ctypes.c_void_p()

status = kc(None, len("keyring_test"), "keyring_test".encode("utf-8"),
            len("foo"), "foo".encode("utf-8"), length, data, None,)

password = ctypes.create_string_buffer(length.value)
ctypes.memmove(password, data.value, length.value)

print("password: {}".format(password.raw.decode("utf-8")))
print("status: {}".format(status))

if status != 0:
    sys.exit(1)
```

When I run this script with a different version of Python, I get the following dialog, as expected:

![screen shot 2016-07-27 at 7 04 48 am](https://cloud.githubusercontent.com/assets/1234956/17176210/d58c5da0-53c8-11e6-8079-98898e8c4ca1.jpg)

I have cleared out all applications from Keychain's `Access Control`, so I expect to get that dialog for every application:

![screen shot 2016-07-27 at 7 10 22 am](https://cloud.githubusercontent.com/assets/1234956/17176305/43b7cb3e-53c9-11e6-9544-a5ee50da4bea.jpg)

When I run the script and click "Allow", I get the following output on working versions of Python:

```bash
$ python2 --version && python2 keychain_ctypes.py 
Python 2.7.12
password: bar
status: 0
```

```bash
$ pyenv/versions/3.5.2/bin/python3 --version
Python 3.5.2
$ pyenv/versions/3.5.2/bin/python3 keychain_ctypes.py 
password: bar
status: 0
```

With the nonworking version (homebrew 3.5.2 from bottle), I don't get the dialog box at all:

```bash
$ python3 --version && python3 keychain_ctypes.py 
Python 3.5.2
password: 
status: -25293
```

I started by removing every trace I can find of homebrew's python3: `brew uninstall python3 && brew cleanup && rm -rf $(brew --cache)`, and afterward going through `find /usr/local -iname '*python3*'` and deleting all relevant files / directories (from e.g. `/usr/local/lib/python3.5`). After that I reinstalled, and nothing changed. `brew doctor` complained about some config scripts from `pyenv`, so I moved those to a directory not on my `PATH`.

Eventually I found that things were still working at 8026463a6723f91f78b4eca946635b32182aa6d2, so I made the little script below and ran a git bisect using my ctypes script from above.

```bash
#! /bin/bash
# test_keyring.sh

brew uninstall python3
rm -rf ~/Library/Caches/Homebrew/python3-*.tar.gz

brew install Formula/python3.rb
/usr/local/bin/python3 ~/fix-keyring/keychain_ctypes.py
status=$?

brew uninstall python3
exit $status
```

```bash
$ git bisect start HEAD 8026463a6723f91f78b4eca946635b32182aa6d2 -- Formula/python3.rb
Bisecting: 14 revisions left to test after this (roughly 4 steps)
[12b677ffdf00465071d86d2c7a728745395a2480] python3: update 3.5.1 bottle.
$ git bisect run bash test_keyring.sh
# ... [lots of output] ...
129021a2525f88fab66065cf2ae2a7a440dcdce1 is the first bad commit
commit 129021a2525f88fab66065cf2ae2a7a440dcdce1
Author: BrewTestBot <brew-test-bot@googlegroups.com>
Date:   Wed Jun 29 13:57:33 2016 +0100

    python3: update 3.5.2 bottle.
```

(I clicked "Always Allow" to avoid having to accept the security dialog every time.)

So inspecting 129021a2525f88fab66065cf2ae2a7a440dcdce1, I can't understand what about it is causing a problem. Thinking that it might be an issue with the bottle, I tried installing with `--build-from-source` **and it works**. 

I've tried reading up on [what bottles are and how they're created](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Bottles.md). So since building from source makes everything work properly, I thought maybe the bottle had been build incorrectly or one of the config flags wasn't right, so I tried building and installing the bottle locally `brew install --build-bottle python3 && brew postinstall python3`, and again, while it looks pretty darn close to the output for building from source, it **doesn't work**:

```bash
$ /usr/local/bin/python3 keychain_ctypes.py
password:
status: -25293
```

The final twist is that this seems to be only affecting my machine -- I find that installing normally (just a regular `brew update && brew upgrade --all && brew install python3`) on both my wife's Macbook Air (virtually identical setup to mine) as well as on a Virtualbox image... both work as expected with my test script.

So does anybody have any suggestions on what is happening differently with the bottle installation that could be causing this issue? Any other debugging steps I should take? I know I can just install from source and go from there, but I'd really like to get to the root of the issue.

**tl;dr:** On my machine, python3 3.5.2 installed from a homebrew bottle can't access the OS X Keychain through the ctypes script above, but it works on other computers (from bottle) as well as on my machine when built from source. What is different about the version installed from a bottle?

Thanks in advance for any help! cc @mikemcquaid [as recommended for bottle issues](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Bottles.md#planned-improvements)

**Comment by tdsmith** [comment]
The most interesting difference is that Homebrew builds with `-march=core2` when bottling, but that doesn't seem super relevant. I have a 2015 i7 MBP and can't reproduce. I don't know what could be causing this.

**Comment by n8henrie** [comment]
Yeah, I tried `brew install --bottle-arch=broadwell python3` and it didn't seem to change anything.

**Comment by n8henrie** [comment]
84897ff215081998c8bc17cc663320ae50f7ccf4 seems to have fixed the issue -- so I wonder if this was related to the "pyport.h macro issue"?

```bash
$ brew uninstall python3
$ git checkout 84897ff215081998c8bc17cc663320ae50f7ccf4
$ brew install --build-bottle Formula/python3.rb
$ /usr/local/bin/python3 keychain_ctypes.py 
password: bar
status: 0
```

Also working with the new bottles (c64ac6c0be2d5957a9a6d3872b94b4a4118b430a)

```bash
$ brew uninstall python3
$ git checkout master
$ brew install Formula/python3.rb
$ /usr/local/bin/python3 keychain_ctypes.py 
password: bar
status: 0
```

Weird. Closing.
